mino/minows/rpc.go

```diff
@@ -0,0 +1,235 @@
+package minows
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/serde"
+	"golang.org/x/xerrors"
+)
+
+const MaxMessageSize = 1e9 // TODO verify
+
+// RPC implements mino.RPC
+// TODO unit tests
+type rpc struct {
+	uri  protocol.ID
+	mino *minows
+	// TODO handler mino.Handler
+	factory serde.Factory
+	context serde.Context
+}
+
+// Call
+// Returns an error if any player address is invalid.
+// Otherwise, returns a response channel 1) filled with replies or errors in
+// the network from each player 2) closed after each player has
+// replied or errored, or the context is done.
+func (r rpc) Call(
+	ctx context.Context,
+	req serde.Message,
+	players mino.Players,
+) (<-chan mino.Response, error) {
+	// TODO assumption: peer store already filled with 'players' Peer IDs & multi-addresses
+	// quit unless all player addresses are valid
+	addrs, err := toAddresses(players)
+	if err != nil {
+		return nil, err
+	}
+	results := openStreams(ctx, r.mino.host, r.uri, addrs)
+	// unicast a request-response to each player concurrently
+	// as streams are established
+	responses := make(chan mino.Response, len(addrs))
+	var wg sync.WaitGroup
+	for range addrs {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			select {
+			case res := <-results: // fan-out
+				if res.err != nil {
+					responses <- mino.NewResponseWithError(res.remote, res.err)
+					return
+				}
+				err := send(res.stream, req, r.context)
+				if err != nil {
+					responses <- mino.NewResponseWithError(res.remote, err)
+					return
+				}
+				reply := receive(res.stream, r.factory, r.context)
+				if err != nil {
+					responses <- mino.NewResponseWithError(reply.sender, reply.err)
+					return
+				}
+				responses <- mino.NewResponse(reply.sender, reply.message)
+			case <-ctx.Done(): // let goroutine exit if context is done
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(responses)
+	}()
+	return responses, nil
+}
+
+// Stream
+// - context defines when the protocol is done,
+// and it should therefore always be canceled at some point. (DELA Doc)
+// - When it's done, all the connections are shut down and the resources are
+// cleaned up (DELA Doc)
+// - orchestrator of a protocol will contact one of the participants which
+// will be the root for the routing algorithm (i.e. gateway?).
+// It will then relay the messages according to the routing algorithm and
+// create relays to other peers when necessary (DELA Doc but ignored)
```

Comment:
```suggestion
// It will then relay the messages according to the routing algorithm, which is
// a direct connection between every pair of nodes in our case.
```